sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,343 @@
+            {
+                usedAs.Add(symbolUsedAs);
+            }
+
+            public bool MatchesIdentifier(IdentifierNameSyntax id, SemanticModel semanticModel)
+            {
+                var symbol = semanticModel.GetSymbolInfo(id).Symbol;
+                return Equals(parameterSymbol, symbol);
+            }
+
+            public Diagnostic GetRuleViolation()
+            {
+                if (ShouldReportOn)
+                {
+                    var mostGeneralType = FindMostGeneralType();
+
+                    if (!Equals(mostGeneralType, parameterSymbol.Type) &&
+                        ShouldReportOnType(mostGeneralType.GetSymbolType()))
+                    {
+                        return Diagnostic.Create(rule,
+                            parameterSymbol.Locations.First(),
+                            mostGeneralType.ToDisplayString(), parameterSymbol.Type.ToDisplayString());
+                    }
+                }
+
+                return null;
+            }
+
+            private static bool ShouldReportOnType(ITypeSymbol typeSymbol)
+            {
+                bool isExcluded =
+                       typeSymbol.Is(KnownType.System_Object) ||
+                       typeSymbol.Is(KnownType.System_ValueType) ||
+                       typeSymbol.Name.StartsWith("_", StringComparison.InvariantCulture) ||
+                       typeSymbol.Is(KnownType.System_Enum);
+
+                return !isExcluded;
+            }
+
+            private ISymbol FindMostGeneralType()
+            {
+                var mostGeneralType = parameterSymbol.Type;
+
+                if (usedAs.Count == 0)
+                {
+                    return mostGeneralType;
+                }
+
+                mostGeneralType = FindMostGeneralClass(mostGeneralType);
+                mostGeneralType = FindMostGeneralInterface(mostGeneralType);
+                return mostGeneralType;
+            }
+
+            private ITypeSymbol FindMostGeneralClass(ITypeSymbol mostGeneralType)
+            {
+                ITypeSymbol currentSymbol = mostGeneralType.BaseType;
+                do
+                {
+                    if (DerivesOrImplementsAll(currentSymbol))
+                    {
+                        mostGeneralType = currentSymbol;
+                    }
+
+                    currentSymbol = currentSymbol?.BaseType;
+                }
+                while (currentSymbol != null);
+
+                return mostGeneralType;
+            }
+
+            private ITypeSymbol FindMostGeneralInterface(ITypeSymbol mostGeneralType, int depth = 0)
+            {
+                if (depth > 10)
+                {
+                    return mostGeneralType;
+                }
+
+                foreach (var iface in mostGeneralType.Interfaces)
+                {
+                    if (DerivesOrImplementsAll(iface))
+                    {
+                        return FindMostGeneralInterface(iface, depth++);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this useless assignment to local variable 'depth'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1854)